model/scope.go

```diff
@@ -0,0 +1,263 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"database/sql/driver"
+	"encoding/json"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+type Scopes []Scope
+type Scope string
+
+const (
+	ScopeChannelJoin  Scope = "channels:join"
+	ScopeMessageRead  Scope = "message:read"
+	ScopeMessageWrite Scope = "message:write"
+	ScopeFilesRead    Scope = "files:read"
+	ScopeFilesWrite   Scope = "files:write"
+	ScopeSearchPosts  Scope = "search:post"
+	ScopeNoScope      Scope = ""
+
+	PluginScopePrefix         = "plugin:"
+	PluginSpecificScopePrefix = "pluginscope:"
+)
+
+func ScopeAny(scopes ...Scope) Scopes {
+	return scopes
+}
+
+func ScopeAllow() Scopes {
```

Comment:
2/5 I still find the representation of Allow as `{}` and Deny as `nil` to be confusing. Is there a strong argument against explicitly codifying `ScopeAllow` as `{""}`?

0/5 I'd add a  `type APIScope Scope` for which `AnyScope = ""` is valid, in addition to all of normal `Scope` values. `nil` for DenyAll seems appropriate, intention-revealing, and the default.

Reply:
> 2/5 I still find the representation of Allow as {} and Deny as nil to be confusing. Is there a strong argument against explicitly codifying ScopeAllow as {""}?

It makes things slightly more complicated in terms of code. The current behavior is simple codewise:
- Nil = deny
- Empty = allow
- With something = scoped

If we make your proposed change, we have way more things to handle:
- Nil = deny
- With something:
  - Has some "" element? If so, it must be of length 1, and means it is allowed. Otherwise is not valid
  - Has no "" element? scoped
- Empty = invalid?

This also may add complexity to other functions, but I haven't gone through it yet.

Reply:
Empty and nil should be equivalent, IMO.

Reply:
From reading the syntax, scopes seems like to form a tree. `files:read` and `files:write` are both leaves it in. Are there also branch nodes? Will there be a scope `file`?

If that is the case, then there is a root leave, it could be named `""`, `":"` or `"."`. Then the metal model stays the same for other scoped cases and for the root: Use the scope to determinate, if the request is valid. If the scope is `root`, it's allowed.

Reply:
The plan is that no, there are no trees. The tree-like names are so they are clearer, and easier to visualize  / organize. We could create some "alias" like `file:all`, but I would shy away from that idea (at least for now).

I am OK using the empty string as any, but just as a "clearer way to do things", not so much as root/leaves mental model.